glommio/src/io/bulk_io.rs

```diff
@@ -417,6 +522,145 @@ impl<U: IoVec + Unpin, S: Stream<Item = (ScheduledSource, U)> + Unpin> Stream
     }
 }
 
+impl<U: IoVec + Unpin, S: Stream<Item = (ScheduledSource, U)> + Unpin> BulkIo<U>
+    for OrderedBulkIo<U, S>
+{
+    fn file(&self) -> &Rc<DmaFile> {
+        &self.file
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct UnorderedBulkIo<U: IoVec + Unpin, S: Stream<Item = (ScheduledSource, U)> + Unpin>
+{
+    file: Rc<DmaFile>,
+    iovs: S,
+
+    inflight: BTreeMap<u64, (ScheduledSource, U)>,
+    concurrency_cap: usize,
+    inflight_memory: usize,
+    memory_cap: usize,
+    terminated: bool,
+
+    waker: Rc<ProxyWaker>,
+}
+
+impl<U: IoVec + Unpin, S: Stream<Item = (ScheduledSource, U)> + Unpin> UnorderedBulkIo<U, S> {
+    pub(crate) fn new(file: Rc<DmaFile>, concurrency: usize, iovs: S) -> UnorderedBulkIo<U, S> {
+        assert!(concurrency > 0);
+        UnorderedBulkIo {
+            file,
+            iovs,
+            inflight: Default::default(),
+            concurrency_cap: concurrency,
+            inflight_memory: 0,
+            memory_cap: usize::MAX,
+            terminated: false,
+            waker: Default::default(),
+        }
+    }
+}
+
+impl<U: IoVec + Unpin, S: Stream<Item = (ScheduledSource, U)> + Unpin> IoConcurrencyLimiter
+    for UnorderedBulkIo<U, S>
+{
+    fn set_concurrency_limit(&mut self, limit: usize) {
+        assert!(limit > 0);
+        assert!(
+            !self.terminated && self.inflight.is_empty(),
+            "should be called before the first call to poll()"
+        );
+        self.concurrency_cap = limit
+    }
+
+    fn set_memory_limit(&mut self, limit: Option<usize>) {
+        assert!(limit.unwrap_or(usize::MAX) > 0);
+        assert!(
+            !self.terminated && self.inflight.is_empty(),
+            "should be called before the first call to poll()"
+        );
+        self.memory_cap = limit.unwrap_or(usize::MAX)
+    }
+
+    fn is_full(&self) -> bool {
+        self.inflight.len() == self.concurrency_cap || self.inflight_memory > self.memory_cap
+    }
+}
+
+impl<U: IoVec + Unpin, S: Stream<Item = (ScheduledSource, U)> + Unpin> Stream
+    for UnorderedBulkIo<U, S>
+{
+    type Item = (ScheduledSource, U);
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        // update the waker
+        this.waker.update(cx.waker().clone());
+
+        // Poll the underlying stream and insert the resulting source, if any, in the
+        // local buffer
+        let poll_inner = |this: &mut Self, cx: &mut Context<'_>| match this.iovs.poll_next(cx) {
+            Poll::Ready(Some(res)) => {
+                this.inflight_memory += res.1.size();
+                let (id, proxy_waker) = this.waker.proxy();
+                res.0.add_waiter_many(proxy_waker);
+                this.inflight.insert(id, res);
```

Comment:
Running into some issues where unordered version only returns 1 ready item.
Might be related to the failing test.  

The waker id that is returned in ``poll_inner`` is always the same,  and the inflight map will keep having only 1 entry inside it and then return ``Poll::Ready(None)`` at a later point.

Currently trying to debug it more.

Reply:
Thanks, @Max-Meldrum! I rebased it this weekend and planned to give it a look today. Let us know here if you find the culprit.

Reply:
Have not gotten around to implementing any fix yet, but the main issue is that were using Rc::as_ptr() as id when there is only a single ProxyWaker instance. that is, always the same pointer address used as id.

```rust
// won't change as its referring to a single instance of Rc<ProxyWaker>
let id = Rc::as_ptr(&rc) as u64;
```